[mlir][ArmSME] Support 4-way widening outer products

c-rhodes opened the pull request:
This patch introduces support for 4-way widening outer products. This enables
the folding of 4 'arm_sme.outerproduct' operations that are chained via the
accumulator into single widened operations.

Changes:

- Adds the following operations:
  - smopa_wide_4way, smops_wide_4way
  - umopa_wide_4way, umops_wide_4way
  - sumopa_wide_4way, sumops_wide_4way
  - sumopa_wide_4way, sumops_wide_4way
- Implements conversions for the above ops to intrinsics in ArmSMEToLLVM.
- Extends 'arm-sme-outer-product' pass.

For a detailed description of these operations see the
'arm_sme.smopa_wide_4way' description.

c-rhodes committed:
[mlir][ArmSME] Support 4-way widening outer products

This patch introduces support for 4-way widening outer products. This enables
the folding of 4 'arm_sme.outerproduct' operations that are chained via the
accumulator into single widened operations.

Changes:

- Adds the following operations:
  - smopa_4way, smops_4way
  - umopa_4way, umops_4way
  - sumopa_4way, sumops_4way
  - sumopa_4way, sumops_4way
- Implements conversions for the above ops to intrinsics in ArmSMEToLLVM.
- Extends 'arm-sme-outer-product' pass.

For a detailed description of these operations see the
'arm_sme.smopa_4way' description.

Address comments. Changes:

- add common match failures.
- move isCompatible to static function.
- update isCompatible to take optional `rhsExtType`.
- use isCompatible in-place of isWidenable.
- add canFuseOuterProducts for 4-way.
- llvm::hasSingleElement -> hasOneUse.
- op.erase -> rewriter.eraseOp.

Address comments. Changes:

Same as 2-way comments.

MacDue commented on mlir/lib/Dialect/ArmSME/Transforms/OuterProductFusion.cpp:
```diff
@@ -225,37 +271,238 @@ class OuterProductFusion2Way
+      return rewriter.notifyMatchFailure(
+          op, MATCH_FAILURE_OUTERPRODUCT_NOT_SINGLE_USE);
+
+    if (bool(op1.getLhsMask()) != bool(op2.getLhsMask()) !=
+        bool(op3.getLhsMask()) != bool(op4.getLhsMask()))
+      return rewriter.notifyMatchFailure(op,
+                                         MATCH_FAILURE_INCONSISTENT_MASKING);
```
This does not make sense for four values (and is very confusing to reason about):

`true != false != false != true` => `false`

c-rhodes replied:
oops, gone now after your suggestion below 👍

MacDue commented on mlir/lib/Dialect/ArmSME/Transforms/OuterProductFusion.cpp:
```diff
@@ -225,37 +271,238 @@ class OuterProductFusion2Way
+    auto nxv2i16 = VectorType::get({2}, rewriter.getI16Type(), true);
+    if (
+        // signed, i8i8i32
+        (failed(
+             isCompatible<arith::ExtSIOp>(rewriter, op1, nxnxv4i32, nxv4i8)) ||
+         failed(
+             isCompatible<arith::ExtSIOp>(rewriter, op2, nxnxv4i32, nxv4i8)) ||
+         failed(
+             isCompatible<arith::ExtSIOp>(rewriter, op3, nxnxv4i32, nxv4i8)) ||
+         failed(
+             isCompatible<arith::ExtSIOp>(rewriter, op4, nxnxv4i32, nxv4i8))) &&
```
The expression:
```c++
        (failed(
             isCompatible<Ext>(rewriter, op1, outputType, inputType)) ||
         failed(
             isCompatible<Ext>(rewriter, op2, outputType, inputType)) ||
         failed(
             isCompatible<Ext>(rewriter, op3, outputType, inputType)) ||
         failed(
             isCompatible<Ext>(rewriter, op4, outputType, inputType)))
```
Really should be moved to lambda (or helper), and reused for each case, it's quite hard to follow all the cases here.

MacDue replied:
A lambda would avoid repeating the arguments again:
```c++
auto failedToMatch = [&](auto extendOp, VectorType resultType, VectorType inputType) {
        using ExtendOpTy = decltype(extendOp);
        return failed(
             isCompatible<ExtendOpTy>(rewriter, op1, resultType, inputType)) ||
         failed(
             isCompatible<ExtendOpTy>(rewriter, op2, resultType, inputType)) ||
         failed(
             isCompatible<ExtendOpTy>(rewriter, op3, resultType, inputType)) ||
         failed(
             isCompatible<ExtendOpTy>(rewriter, op4, resultType, inputType)));
};
```

// using:
```c++
failedToMatch(arith::ExtUIOp{}, nxnxv2i64, nxv2i16)
```

(passing the null `arith::ExtUIOp{}` is a trick to use it as an implicit template parameter, since template proper template lambdas need C++20)

c-rhodes replied:
this is much better cheers, my c++ is pretty woeful.. 😅

MacDue reviewed (commented):
Few initial comments:

c-rhodes committed:
Address comments. Changes:

- fix check for consistent masking.
- rewrite as loop that walks outer product chain.
- use lambda for match check.

c-rhodes commented on mlir/lib/Dialect/ArmSME/Transforms/OuterProductFusion.cpp:
```diff
@@ -225,37 +271,238 @@ class OuterProductFusion2Way
+    if (bool(op1.getLhsMask()) != bool(op2.getLhsMask()) !=
+        bool(op3.getLhsMask()) != bool(op4.getLhsMask()))
+      return rewriter.notifyMatchFailure(op,
+                                         MATCH_FAILURE_INCONSISTENT_MASKING);
```
this is a lot cleaner, cheers!

banach-space commented on mlir/lib/Dialect/ArmSME/Transforms/OuterProductFusion.cpp:
```diff
@@ -225,37 +271,173 @@ class OuterProductFusion2Way
+    auto rhsExtOp = op.getRhs().getDefiningOp();
+
+    arm_sme::CombiningKind kind = op.getKind();
+    if (kind == arm_sme::CombiningKind::Add) {
+      if (isa<arith::ExtSIOp>(lhsExtOp) && isa<arith::ExtSIOp>(rhsExtOp))
+        rewriter.replaceOpWithNewOp<arm_sme::SMopa4WayOp>(
+            op4, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
+      else if (isa<arith::ExtUIOp>(lhsExtOp) && isa<arith::ExtUIOp>(rhsExtOp))
+        rewriter.replaceOpWithNewOp<arm_sme::UMopa4WayOp>(
+            op4, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
+      else if (isa<arith::ExtSIOp>(lhsExtOp) && isa<arith::ExtUIOp>(rhsExtOp))
+        rewriter.replaceOpWithNewOp<arm_sme::SuMopa4WayOp>(
+            op4, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
+      else if (isa<arith::ExtUIOp>(lhsExtOp) && isa<arith::ExtSIOp>(rhsExtOp))
+        rewriter.replaceOpWithNewOp<arm_sme::UsMopa4WayOp>(
+            op4, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
+      else
+        llvm_unreachable("unexpected extend op!");
+    } else if (kind == arm_sme::CombiningKind::Sub) {
+      if (isa<arith::ExtSIOp>(lhsExtOp) && isa<arith::ExtSIOp>(rhsExtOp))
+        rewriter.replaceOpWithNewOp<arm_sme::SMops4WayOp>(
+            op4, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
+      else if (isa<arith::ExtUIOp>(lhsExtOp) && isa<arith::ExtUIOp>(rhsExtOp))
+        rewriter.replaceOpWithNewOp<arm_sme::UMops4WayOp>(
+            op4, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
+      else if (isa<arith::ExtSIOp>(lhsExtOp) && isa<arith::ExtUIOp>(rhsExtOp))
+        rewriter.replaceOpWithNewOp<arm_sme::SuMops4WayOp>(
+            op4, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
+      else if (isa<arith::ExtUIOp>(lhsExtOp) && isa<arith::ExtSIOp>(rhsExtOp))
+        rewriter.replaceOpWithNewOp<arm_sme::UsMops4WayOp>(
+            op4, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
+      else
+        llvm_unreachable("unexpected extend op!");
+    } else {
+      llvm_unreachable("unexpected arm_sme::CombiningKind!");
+    }
```
Use `TypeSwitch`? Similarly to 2-way ;-)

c-rhodes replied:
there's two types to switch on because of the signed-by-unsigned / unsigned-by-signed variants and i don't believe it's possible to use `TypeSwitch` for this

banach-space replied:
Missed that! Not sure how to improve this then :/

MacDue commented on mlir/lib/Dialect/ArmSME/Transforms/OuterProductFusion.cpp:
```diff
@@ -225,37 +271,173 @@ class OuterProductFusion2Way
+      else if (isa<arith::ExtSIOp>(lhsExtOp) && isa<arith::ExtUIOp>(rhsExtOp))
+        rewriter.replaceOpWithNewOp<arm_sme::SuMopa4WayOp>(
+            op4, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
+      else if (isa<arith::ExtUIOp>(lhsExtOp) && isa<arith::ExtSIOp>(rhsExtOp))
+        rewriter.replaceOpWithNewOp<arm_sme::UsMopa4WayOp>(
+            op4, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
```
Why do mixed zero-extended and sign-extened inputs become a unsigned mopa?

c-rhodes replied:
> Why do mixed zero-extended and sign-extened inputs become a unsigned mopa?

this is a unsigned-by-signed mopa

MacDue replied:
Needs some comments :) The difference in names is barely noticeable (and not clear without looking at the definitions)

c-rhodes committed:
Address comments

add comment to clarify each variant.

c-rhodes committed:
Rebase and fix op descriptions

c-rhodes committed:
Fix broken doc links to smopa_4way

banach-space commented on mlir/test/Dialect/ArmSME/outer-product-fusion.mlir:
```diff
@@ -291,6 +894,41 @@ func.func @outerproduct_widening_2way__bad_acc(%a0 : vector<[4]xf16>, %b0 : vect
 
 // -----
 
+// CHECK-LABEL: @outerproduct_widening_4way__bad_acc
```
```suggestion
// CHECK-LABEL: @outerproduct_widening_4way__missing_acc
```

banach-space commented on mlir/test/Dialect/ArmSME/outer-product-fusion.mlir:
```diff
@@ -291,6 +894,41 @@ func.func @outerproduct_widening_2way__bad_acc(%a0 : vector<[4]xf16>, %b0 : vect
+  %0 = arm_sme.outerproduct %a0_ext, %b0_ext : vector<[4]xi32>, vector<[4]xi32>
+  %1 = arm_sme.outerproduct %a1_ext, %b1_ext acc(%0) : vector<[4]xi32>, vector<[4]xi32>
+  %2 = arm_sme.outerproduct %a2_ext, %b2_ext acc(%1) : vector<[4]xi32>, vector<[4]xi32>
+  // break chain
```
```suggestion
  // Missing accumulate - breaks chain of use-def
```

banach-space commented on mlir/test/Dialect/ArmSME/outer-product-fusion.mlir:
```diff
@@ -314,6 +952,40 @@ func.func @outerproduct_widening_2way__bad_combining_kind(
 
 // -----
 
+// CHECK-LABEL: @outerproduct_widening_4way__bad_combining_kind
```
```suggestion
// CHECK-LABEL: @outerproduct_widening_4way__inconsistent_combining_kind
```

banach-space commented on mlir/test/Dialect/ArmSME/outer-product-fusion.mlir:
```diff
@@ -344,6 +1016,41 @@ func.func @outerproduct_widening_2way__cant_erase(
 
 // -----
 
+// CHECK-LABEL: @outerproduct_widening_4way__cant_erase
```
```suggestion
// CHECK-LABEL: @outerproduct_widening_4way__multiple_use
```

banach-space commented on mlir/test/Dialect/ArmSME/outer-product-fusion.mlir:
```diff
@@ -389,6 +1130,41 @@ func.func @outerproduct_widening_2way__bad_masking(
 
 // -----
 
+// CHECK-LABEL: @outerproduct_widening_4way__bad_masking
```
```suggestion
// CHECK-LABEL: @outerproduct_widening_4way__inconsistent_masking
```

banach-space commented on mlir/test/Dialect/ArmSME/outer-product-fusion.mlir:
```diff
@@ -405,6 +1181,37 @@ func.func @outerproduct_widening_2way__bad_defining_op(
 
+// -----
+
+// CHECK-LABEL: @outerproduct_widening_4way__bad_defining_op
```
Not sure I see what's "wrong" with this example :)

c-rhodes replied:
same as above but for 4-way, i.e. Defining op of outer product must be a supported extension op.

But I've added a comment where it goes wrong to clarify.

banach-space commented on mlir/test/Integration/Dialect/Vector/CPU/ArmSME/test-outerproduct-i8i8i32.mlir:
```diff
@@ -0,0 +1,150 @@
+  %c2 = arith.constant 2 : index
+  %c3 = arith.constant 3 : index
+  %c4 = arith.constant 4 : index
+
```
[nit] Add a comment that with these masks the bottom row of the output matrix is preserved

c-rhodes replied:
I've not added this one as that isn't all they do and I think it could be misleading.

banach-space commented on mlir/lib/Dialect/ArmSME/Transforms/OuterProductFusion.cpp:
```diff
@@ -31,6 +31,55 @@ using namespace mlir;
+
+// Common match failure reasons.
+static constexpr StringLiteral
+    MATCH_FAILURE_NO_ACCUMULATOR("no accumulator operand");
```
Aren't ALL_CAPS reserved for macros? Shouldn't these be camelCase? https://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly

MacDue replied:
Just an aside/observation: MLIR is not following the LLVM coding standards exactly, it's a mish-mash of Google style + LLVM style (but follows neither exactly), in both new and old code.

MacDue replied:
E.g. MLIR is not following the LLVM rules for variable names (nor the Google one for that matter). But it does seem to follow Google constant naming rules.

c-rhodes replied:
>Just an aside/observation: MLIR is not following the LLVM coding standards exactly, it's a mish-mash of Google style + LLVM style (but follows neither exactly), in both new and old code.

Yeah I noticed in that doc it says variables start with upper case but that's not true in MLIR. Anyhow, for this particular case, I can't see anywhere else that uses upper case for constexpr like this so I'll change to camel case given it's prevalent.

banach-space reviewed (approved):
Some small nits/suggestions, otherwise LGTM, thanks!

c-rhodes committed:
Address comments

c-rhodes committed:
Address comments

MacDue reviewed (approved):
LGTM

MacDue commented on mlir/lib/Dialect/ArmSME/Transforms/OuterProductFusion.cpp:
```diff
@@ -31,6 +31,65 @@ using namespace mlir;
+
+// Common match failure reasons.
+static constexpr StringLiteral
+    matchFailureNoAccumulator("no accumulator operand");
+static constexpr StringLiteral matchFailureExpectedOuterProductDefOp(
+    "defining op of accumulator must be 'arm_sme.outerproduct'");
+static constexpr StringLiteral matchFailureInconsistentCombiningKind(
+    "combining kind (add or sub) of outer products must match");
+static constexpr StringLiteral matchFailureInconsistentMasking(
+    "unsupported masking, either both outerproducts are masked "
+    "or neither");
+static constexpr StringLiteral matchFailureOuterProductNotSingleUse(
+    "outer product(s) not single use and cannot be removed, no benefit to "
```
very nit: It seems more common to prefix constants with `k`, so `kMatch...`

banach-space replied:
+1

c-rhodes committed:
Address comments
